consolidation/consolidate.go

```diff
@@ -59,3 +59,69 @@ func AggEvery(numPoints, maxPoints uint32) uint32 {
+	}
+	points = Consolidate(points, aggNum, consolidator)
+	interval *= aggNum
+	return points, interval
+}
+
+// Nudge computes the parameters for nudging.
+// let's say a series has points A,B,C,D and we must consolidate with numAgg=2.
+// if we wait a step, point E appears into the visible window and A will slide out of the window.
+// there's a few approaches you can take wrt such changes across refreshes:
+// 1) naive old approach:
+//    on first load return consolidate(A,B), consolidate(C,D)
+//    after a step, return consolidate(B,C), consolidate(D,E)
+//    => this looks weird across refreshes:
+//       both the values as well as the timestamps change everywhere, points jump around on the chart
+// 2) graphite-style nudging: trim a few of the first points away as needed, so that the first TS
+//    is always a multiple of the postConsolidationInterval (note: assumes input is quantized!)
+//    on first load return consolidate(A,B), consolidate(C,D)
+//    after a step, return consolidate(C,D), consolidate(E)
+//    => same points are always consolidated together, no jumping around.
+//    => simple to understand, but not the most performant (fetches/processes some points needlessly)
+//    => tends to introduce emptyness in graphs right after the requested start.
+//       (because Grafana plots from requested start, not returned start, and so there will be some empty space
+//       where we trimmed points)
+// 3) similar to 2, but don't trim, rather consolidate the leftovers both on the start and at the end.
+//    on first load return consolidate(A,B), consolidate(C,D)
+//    after a step, return consolidate(B), consolidate(C,D), consolidate(E)
+//    => same points are always consolidated together, no jumping around.
+//    => only datapoint up front and at the end may jump around, but not the vast majority
+//    => no discarding of points
+//    => requires a large code change though, as it makes it harder to honor MaxDataPoints.
+//       e.g. MDP=1, you have 5 points and aggNum is 5, if alignment is improper, it would cause
+//       2 output points, so we would have to rewrite a lot of code, no longer compute AggNum in advance etc
+//
+// note that with all 3 approaches, we always consolidate leftovers at the end together, so with any approach
+// the last point may jump around (see Consolidate function)
+// for now, and for simplicity we just implement the 2nd approach. it's also the only one that assures MDP is strictly
+// honored (see last point of approach 3, which also affects approach 1)
+func nudge(start, preAggInterval, aggNum uint32) (uint32, int) {
+	postAggInterval := preAggInterval * aggNum
+	var num int
+	var diff uint32
+	// move start until it maps to the first point of an aggregation bucket
+	// since clean multiples of the new postAggInterval are the last point to go into an aggregation
+	// we want a point that comes preAggInterval after it.
+	remainder := (start - preAggInterval) % postAggInterval
```

Comment:
I have pushed a commit which should explain this stuff (and some of the other stuff) better. let me know what you think / if that clears it up.